invenio_deposit/api.py

```diff
@@ -210,52 +260,10 @@ def publish(self, pid=None, id_=None):
         self['_deposit']['status'] = 'published'
 
         if self['_deposit'].get('pid') is None:  # First publishing
-            minter = current_pidstore.minters[
-                current_app.config['DEPOSIT_PID_MINTER']
-            ]
-            id_ = id_ or uuid.uuid4()
-            record_pid = minter(id_, self)
-
-            self['_deposit']['pid'] = {
-                'type': record_pid.pid_type, 'value': record_pid.pid_value,
-                'revision_id': 0,
-            }
-
-            data = dict(self.dumps())
-            data['$schema'] = self.record_schema
-
-            # During first publishing create snapshot the bucket.
-            @contextmanager
-            def process_files(data):
-                """Process deposit files."""
-                if RecordsBuckets.query.filter_by(
-                        record_id=self.model.id).count():
-                    assert not self.files.bucket.locked
-                    self.files.bucket.locked = True
-                    snapshot = self.files.bucket.snapshot(lock=True)
-                    data['_files'] = self.files.dumps(bucket=snapshot.id)
-                    yield data
-                    db.session.add(RecordsBuckets(
-                        record_id=id_, bucket_id=snapshot.id
-                    ))
-                else:
-                    yield data
-
-            with process_files(data) as data:
-                record = Record.create(data, id_=id_)
-
+            self._publish_new(id_=id_)
```

Comment:
It could then require four such "signals": ``pre/post_publish/edit``: https://github.com/krzysztof/zenodo/blob/communities_merging/zenodo/modules/deposit/api.py#L102